subsys/usb/device_next/class/usbd_cdc_ecm.c

```diff
@@ -0,0 +1,724 @@
+	LOG_DBG("Received packet len %zu", net_pkt_get_len(pkt));
+	if (net_recv_data(data->iface, pkt) < 0) {
+		LOG_ERR("Packet %p dropped by network stack", pkt);
+		net_pkt_unref(pkt);
+	}
+
+restart_out_transfer:
+	net_buf_unref(buf);
+	atomic_clear_bit(&data->state, CDC_ECM_OUT_ENGAGED);
+
+	return cdc_ecm_out_start(c_nd);
+}
+
+static int usbd_cdc_ecm_request(struct usbd_class_node *const c_nd,
+				struct net_buf *buf, int err)
+{
+	struct usbd_contex *uds_ctx = c_nd->data->uds_ctx;
+	const struct device *dev = c_nd->data->priv;
+	struct cdc_ecm_eth_data *data = dev->data;
+	struct udc_buf_info *bi;
+
+	bi = udc_get_buf_info(buf);
+
+	if (bi->ep == cdc_ecm_get_bulk_out(c_nd)) {
+		return cdc_ecm_acl_out_cb(c_nd, buf, err);
+	}
+
+	if (bi->ep == cdc_ecm_get_bulk_in(c_nd)) {
+		k_sem_give(&data->sync_sem);
+
+		return 0;
+	}
+
+	if (bi->ep == cdc_ecm_get_int_in(c_nd)) {
+		k_sem_give(&data->notif_sem);
+
+		return 0;
+	}
+
+	return usbd_ep_buf_free(uds_ctx, buf);
+}
+
+static int cdc_ecm_send_notification(const struct device *dev,
+				     const bool connected)
+{
+	struct cdc_ecm_eth_data *data = dev->data;
+	struct usbd_class_node *c_nd = data->c_nd;
+	struct cdc_ecm_notification notification = {
+		.RequestType = {
+			.direction = USB_REQTYPE_DIR_TO_HOST,
+			.type = USB_REQTYPE_TYPE_CLASS,
+			.recipient = USB_REQTYPE_RECIPIENT_INTERFACE,
+		},
+		.bNotificationType = USB_CDC_NETWORK_CONNECTION,
+		.wValue = sys_cpu_to_le16((uint16_t)connected),
+		.wIndex = sys_cpu_to_le16(cdc_ecm_get_ctrl_if(c_nd)),
+		.wLength = 0,
+	};
+	struct net_buf *buf;
+	uint8_t ep;
+	int ret;
+
+	if (!atomic_test_bit(&data->state, CDC_ECM_CLASS_ENABLED)) {
+		LOG_INF("USB configuration is not enabled");
+		return 0;
+	}
+
+	if (atomic_test_bit(&data->state, CDC_ECM_CLASS_SUSPENDED)) {
+		LOG_INF("USB support is suspended (FIXME)");
+		return 0;
+	}
+
+	ep = cdc_ecm_get_int_in(c_nd);
+	buf = usbd_ep_buf_alloc(c_nd, ep, sizeof(struct cdc_ecm_notification));
+	if (buf == NULL) {
+		return -ENOMEM;
+	}
+
+	net_buf_add_mem(buf, &notification, sizeof(struct cdc_ecm_notification));
+	ret = usbd_ep_enqueue(c_nd, buf);
+	if (ret) {
+		LOG_ERR("Failed to enqueue net_buf for 0x%02x", ep);
+		net_buf_unref(buf);
+		return ret;
+	}
+
+	k_sem_take(&data->notif_sem, K_FOREVER);
+	net_buf_unref(buf);
+
+	return 0;
+}
+
+static void usbd_cdc_ecm_update(struct usbd_class_node *const c_nd,
+				const uint8_t iface, const uint8_t alternate)
+{
+	struct usbd_cdc_ecm_desc *desc = c_nd->data->desc;
+	const uint8_t data_iface = desc->if1_1.bInterfaceNumber;
+	const struct device *dev = c_nd->data->priv;
+	struct cdc_ecm_eth_data *data = dev->data;
+
+	LOG_INF("New configuration, interface %u alternate %u",
+		iface, alternate);
+
+	if (data_iface == iface && alternate == 0) {
+		net_if_carrier_off(data->iface);
+	}
+
+	if (data_iface == iface && alternate == 1) {
+		net_if_carrier_on(data->iface);
+		if (cdc_ecm_out_start(c_nd)) {
+			LOG_ERR("Failed to start OUT transfer");
+		}
+
+	}
```

Comment:
Minor, could be simplified by factoring out :
```
if (data_iface != iface) {
	return;
}
```
then `else if` for the `alternate`.

Reply:
```suggestion
	if (data_iface != iface) {
		return;
	}

	if (alternate == 0) {
 		net_if_carrier_off(data->iface);
 	} else if (alternate == 1) {
		net_if_carrier_on(data->iface);
		if (cdc_ecm_out_start(c_nd)) {
			LOG_ERR("Failed to start OUT transfer");
		}
	}
```

Reply:
It simplifies nothing.